### PR TITLE
Rectifier polls singularity for deploy to complete.

### PR DESCRIPTION
**Pull request by nyarly**
This'll mean that the time between Rectification.Begin() is called and Wait() will
return will be much longer, but once it completes, the service should either be
live in Singularity or failed.

Room for future work: synchronzing the Rectification.Resolution and reporting
it on the endpoint to give "live-er" responses to users.

### CODE REVIEW

**Comment by sous-machine-account** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR: 635 lines of code changed</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by lachlanh** [comment]
big pr, understand some of it, can see where it is making singularity requests, can't identify the polling mechanism. would probably be good to get another set of eyes on this.

File: lib/rectification.go [hunk 1]
```diff
@@ -1,42 +1,107 @@
...
+			clusters = rf.FilteredClusters(clusters)
+
+			// TODO constants / configs
+			tick := time.NewTicker(250 * time.Millisecond)
+			end, ec := context.WithTimeout(r.ctx, 20*time.Minute)
+			defer ec()
+
+			defer tick.Stop()
+			for {
+				r.pollOnce(d, reg, clusters)
```
Comment (nyarly) [inline lib/rectification.go]: @lachlanh Polling occurs around here.

File: ext/singularity/deployment_builder_test.go [hunk 2]
```diff
@@ -147,39 +136,39 @@ func TestBuildDeployment(t *testing.T) {
 
 	log, _ := logging.NewLogSinkSpy()
 
-	fakeSing := &fakeSingClient{
-		cannedAnswer: &dtos.SingularityDeployHistory{
-			DeployResult: &dtos.SingularityDeployResult{
-				DeployState: dtos.SingularityDeployResultDeployStateSUCCEEDED,
+	cannedDep := &dtos.SingularityDeployHistory{
+		DeployResult: &dtos.SingularityDeployResult{
+			DeployState: dtos.SingularityDeployResultDeployStateSUCCEEDED,
```
Comment (JC1738) [inline ext/singularity/deployment_builder_test.go]: That's a mouthful

File: ext/singularity/requeststate.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	clusterName := pair.Post.Deployment.ClusterName
+	if cluster, has := clusters[clusterName]; has {
+		url = cluster.BaseURL
+	} else {
+		return nil, errors.Errorf("No cluster found for %q. Known are: %q.", clusterName, clusters.Names())
+	}
+
+	client := r.buildSingClient(url)
+
+	reqParent, err := client.GetRequest(reqID, false) //don't use the web cache
```
Comment (JC1738) [inline ext/singularity/requeststate.go]: what's the web cache?

File: ext/singularity/singclient.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
+package singularity
```
Comment (JC1738) [inline ext/singularity/singclient.go]: What's the naming convention, this is a test file, but packaged in etc/singularity?

File: vendor/github.com/samsalisbury/psyringe/psyringe.go [hunk 5]
```diff
@@ -66,6 +66,7 @@ type Psyringe struct {
 	scope          string
 	injectionTypes injectionTypes
 	Hooks          Hooks
+	allowAddCycle  bool
```
Comment (JC1738) [inline vendor/github.com/samsalisbury/psyringe/psyringe.go]: was this used in this PR?

**Comment by JC1738** [comment]
Worried about the time it takes, but don't think should block PR or further testing.

File: vendor/github.com/samsalisbury/psyringe/psyringe.go (hunk 5 above)
Comment (samsalisbury) [reply vendor/github.com/samsalisbury/psyringe/psyringe.go]: No, there's no way to set that to true from outside the package. This flag exists only for internal testing and running comparative benchmarks, specifically https://github.com/samsalisbury/psyringe/blob/master/psyringe_bench_test.go; will likely be removed entirely soon.

**Pull request merged**